store/storage/db.go

```diff
@@ -0,0 +1,221 @@
+package storage
+
+import (
+	"fmt"
+	"sync"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"cosmossdk.io/store/v2"
+)
+
+// TODO:
+//
+// - Define and implement pruning method(s)
+// - Import/snapshot method(s)???
+// - Should we make this an interface?
+//
+// Ref: https://github.com/cosmos/cosmos-sdk/issues/17279
+
+// DefaultMemDBSize defines the default pre-allocation size of the in-memory
+// database used to accumulate database entries prior to committing them to disk.
+const DefaultMemDBSize = 1000
+
+var (
+	_ store.VersionedReaderWriter    = (*Database)(nil)
+	_ store.VersionedIteratorCreator = (*Database)(nil)
+	_ store.Committer                = (*Database)(nil)
+)
+
+// Database defines the state storage (SS) backend. Each open instance of a Database
+// is meant to commit key/value pairs to disk upon Commit for a single version.
+// However, it can be used to read key/value pairs at any version.
+type Database struct {
+	mu      sync.RWMutex
```

Comment:
We discourage using sync.RWMutex as you really can't predict right now the ratio of reads to writes, yet sync.RWMutex scales poorly. Please take a look at our recommendation https://cyber.orijtech.com/scsec/cosmos-go-coding-guide#avoid-syncrwmutex